### PR TITLE
Precompile for Poseidon2 BabyBear

### PR DESCRIPTION
**Pull request by lvella**

### CODE REVIEW

File: std/machines/hash/poseidon2_bb.asm
Comment (chriseth) [inline std/machines/hash/poseidon2_bb.asm:166]: Can this be shared across the different poseidon implementations?

File: std/machines/hash/poseidon2_bb.asm
Comment (chriseth) [inline std/machines/hash/poseidon2_bb.asm:128]: Can these be shared across the different poseidon implementations?

Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:128]: No, because the number of rounds depends on the state length, which depends on the field size. Also these constants don't fit in BabyBear.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:166]: No, because the power you use for the S-Box depends on the field. But 7 is used both by BabyBear and Goldilocks.

File: std/machines/hash/poseidon2_bb.asm
Comment (chriseth) [inline std/machines/hash/poseidon2_bb.asm:56]: Could we please avoid copying these magic constants across 4-5 files?

Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:56]: They are not. The only other place most of these constants appear is inside plonky3 rust code.

The best I can do is, after merging this branch with the poseidon_bb one, place a fraction of the big list of random numbers that is common between the two of them in a separate file, as they are the output of the same deterministic RNG.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm]: Done.

**Comment by lvella** [comment]
Blocked on #1933.

File: std/machines/hash/poseidon2_bb.asm
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:18]:
```suggestion
// As it stands, it cannot be used in a Sponge construction, because we don't
```

File: std/machines/hash/poseidon2_bb.asm
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:58]: can you make an issue for this?

File: std/machines/hash/poseidon2_bb.asm
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:197]:
```suggestion
    // TODO: when link is available inside functions, we can turn this into array operations.
```

File: std/machines/hash/poseidon2_bb.asm
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:198]: @georgwiese all these links :grimacing:

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:58]: I am not even sure we want it.

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:192]: I think this is for another time, but I wonder if there is a way not to commit to these values. Couldn't `increment_ptr ` return expressions for the incremented address instead of constraining an witness column to be equal to it? Not sure what happens when you pass that into `increment_ptr` again though, maybe you'd get a high-degree expression?

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:41]:
```suggestion
    let latch = 1;
```
This works too and does not create a fixed column!

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:198]: Don't be afraid of links ;)

In the optimized case, this should create 2 columns + some constraints for each link. But the whole machine has block size 1, so it will still be few cells to commit to. I think the main disadvantage of this approach vs larger block size & fewer links is increased verifier time (because it scales with the number of columns & constraints). We'll have to see what the tradeoff space is here.

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:237]: Again, we should optimize this later, but do you think we have to commit to all the intermediate states? Could `{internal,external}_round()` return an expression and we just use that?

Since Poseidon is an algebraic hash function, I think we could in principle express the entire function as one big algebraic expression. The only problem would be that its would be of too high degree. But eventually we should try to minimize the witness columns.

**Review by georgwiese — COMMENTED** [review]
This looks nice! :) I have a few questions, but I think many of my suggestions could be implemented in future PRs.

File: plonky3/src/bin/gen_poseidon2_bb_consts.rs
Comment (lvella) [reply plonky3/src/bin/gen_poseidon2_bb_consts.rs]: fixed

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:16]: > Hmm, so the function permutes a 512-Bit input and returns the first 256 Bit, right?

Right. This is the version used by p3 backend, so it should be useful for recursion.

> This is different from the GL Poseidon, which permutes a 768-Bit input (12 64-Bit field elements) and returns the first 256 Bit. There, you'd input your 512-Bit input (e.g. the two child hashes in a Merkle tree) plus a 256-Bit capacity, which would be a constant (where different constants give you different hash functions).

Yes.

> Are you sure we're not supposed to put in a 256-Bit constant here as well? How would we get different hash functions (e.g. one for the Merkle leaves and one for the internal nodes)?

For internal nodes a full hash function is not necessary, just a compression is sufficient. Which this is.

For the leaves, I don't know for sure. How do you do with GL?

I can see 2 options. Alternative a) is Merkle-tree all the way down to 512 bits of data (it works because the data is a fixed power of 2 size). I guess this is safe because this is an intended usage in Poseidon2 paper.

Alternative b) is to use a construction that can turn a compression function into a hash function, like Merkle–Damgård (or its updated version, HAIFA construct). This usage is just implied in passing in Poseidon2 paper. If you need 2^256 different hash functions, set the IV to something other than zero in the first compression step, but I don't see why it would be needed in the bootloader.

Alternative a) seems simpler to me.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:147]: Notice how the "AddRoundConstants(·)" is still applied to the entire state vector in this figure, not only the s-boxed element. In the p3 version, a constant is only added to state[0]. I am not sure how this affects security.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm]: Added comment explaining the matrix and how the multiplication works.

The whole point of Poseidon2 is that this multiplication can be heavily optimized.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:218]: I am pretty sure it would extrapolate the degree limit, due to the logic to juggle the carry between the two limbs.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:263]: Yes.

See the issue I opened related to this: #1971

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:217]: I believe this was my first attempt. Let me try again.

Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:217]: Nope.

```
Cannot unify types () and std::prelude::Constr
    ┌─ /home/lucas/powdr/src/powdr/std/machines/hash/poseidon2_bb.asm:208:13
    │  
208 │ ╭             array::zip(
209 │ │                 array::sub_array(addr, 0, array::len(addr) - 1),
210 │ │                 array::sub_array(addr, 1, array::len(addr) - 1),
211 │ │                 constr |(high, low), (next_high, next_low)| {
212 │ │                     increment_ptr(4, high, low, next_high, next_low);
213 │ │                 }
214 │ │             );
    │ ╰─────────────^

thread 'poseidon2_bb_test' panicked at /home/lucas/powdr/src/powdr/pipeline/src/test_util.rs:325:51:
called `Result::unwrap()` on an `Err` value: ["Expected type std::prelude::Constr[] but got type ()[].\nCannot unify types () and std::prelude::Constr at /home/lucas/powdr/src/powdr/std/machines/hash/poseidon2_bb.asm:9681-9986"]
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm]: Done.

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:267]: Actually, not really. If you multiply 2 polynomials, the resulting polynomial has size m*n, so they can grow to intractable sizes.

Unless we internally represent polynomials as a DAG of multiplication of polynomial factors (which I guess neither we nor the backends do).

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:16]: > For internal nodes a full hash function is not necessary, just a compression is sufficient. Which this is.

What do you mean by "compression is sufficient"? I think we need collision resistance, right? I would *guess* that this construction would be collision resistant, I'm just wondering if that's a standard implementation or whether we do something unusual here.

And yes, for the bootloader the Merkle tree in the bootloader has a pre-defined depth, so it is fine to use the same hash function!

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:267]: > If you multiply 2 polynomials, the resulting polynomial has size m*n, so they can grow to intractable sizes.

I don't understand this comment! If you multiply two polynomials of degrees m and n, you get a polynomial of degree m + n, right?

I do think that #1971 would fix this: We'd add each intermediate state as an intermediate polynomials, which depending on the degree bound would be inlined or materialized as a witness column.

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:186]: ❤️

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:154]: Right! Strange indeed, but I guess we want to by consistent with Plonky3.

**Review by georgwiese — APPROVED** [review]
Cool!

File: std/machines/hash/poseidon2_bb.asm
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:267]: Oh, maybe you're talking about the fact that if e.g. if you repeatedly compute an S-Box $x^7$ and $x$ is an expression and you inline everything, the resulting expression grows exponentially? That's fixed by intermediate columns and Thibaut just pushed #1995, which makes the Plonky3 backend aware of them :)

File: std/machines/hash/poseidon2_bb.asm
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:16]: This is a quote from Poseidon2 paper that I interpret as meaning it is safe to use a single Poseidon2 permutation to level-up a Merkle-tree:

> While hash functions are perfectly usable and allow to make strong security
arguments, they are often more generic. For many classical applications (e.g.,
ordinary hashing of arbitrary-length inputs) this is not a major issue. However,
when building a Merkle tree with small fixed-size input lengths, we often only
need a single permutation call to process the entire input and compute the
desired output. In this case, the construction is not used as a hash function in a
traditional sense, but rather as a compression function.

**Pull request merged**
